crates/cubecl-reduce/src/shared.rs

```diff
@@ -0,0 +1,408 @@
+
+    // Write the first item of accumulator into the ouput at the given index.
+    fn write_first<EO: Numeric>(
+        output: &mut Tensor<Line<EO>>,
+        accumulator: &Self::Accumulator,
+        index: u32,
+        shape_reduce_dim: u32,
+    );
+}
+
+#[cube]
+pub fn reduce_shared<RD: ReduceSharedInstruction<EI>, EI: Numeric, EO: Numeric>(
+    input: &Tensor<Line<EI>>,
+    output: &mut Tensor<Line<EO>>,
+    reduce_dim: u32,
+    #[comptime] cube_dim: u32,
+    #[comptime] exact_shape: bool,
+) {
+    let line_size = input.line_size();
+    let mut accumulator = RD::create_accumulator(cube_dim, line_size);
+    RD::init_accumulator(&mut accumulator, UNIT_POS, line_size);
+
+    // Compute the first index where to start the reduction for the current cube.
+    // First, compute the coordinate corresponding to the CUBE_POS element of the output tensor
+    // Then, use the strides of the input tensor to find the index of the same coordinate
+    // in the input tensor.
+    let mut offset = 0;
+    for axis in 0..input.rank() {
+        let coordinate = output.coordinate(CUBE_POS, axis);
+        offset += coordinate * input.stride(axis);
+    }
+
+    let num_items_per_unit = if exact_shape {
+        input.shape(reduce_dim) / cube_dim
+    } else {
+        div_ceil(input.shape(reduce_dim), cube_dim)
+    };
+
+    // The unit at UNIT_POS reduces the items with coordinate (UNIT_POS, UNIT_POS + cube_dim, UNIT_POS + 2 * cube_dim, ...).
+    // The result in stored into the accumulator at index UNIT_POS.
+    for i in 0..num_items_per_unit {
+        let coordinate = i * cube_dim + UNIT_POS;
+        let index = offset + coordinate * input.stride(reduce_dim);
+        #[allow(clippy::collapsible_else_if)]
+        if exact_shape {
+            RD::accumulate(&mut accumulator, UNIT_POS, input[index], coordinate);
+        } else {
+            if coordinate < input.shape(reduce_dim) {
+                RD::accumulate(&mut accumulator, UNIT_POS, input[index], coordinate);
+            }
+        }
+    }
+    sync_units();
+
+    // Merge the accumulator like this with some padding if CUBE_DIM is not a power of 2.
+    //
+    // 0   1   2   3   4   5   6   7
+    // |   |   |   |   |   |   |   |
+    // +-+-+   +-+-+   +-+-+   +-+-+
+    //   |       |       |       |
+    //   +---+---+       +---+---+
+    //       |               |
+    //       +-------+-------+
+    //               |
+    //               *
+    if comptime!(cube_dim.is_power_of_two()) {
+        let mut num_active_units = cube_dim;
+        let mut jump = 1;
+        while num_active_units > 1 {
+            num_active_units /= 2;
+            let destination = jump * 2 * UNIT_POS;
+            let origin = jump * (2 * UNIT_POS + 1);
+            if UNIT_POS < num_active_units {
+                RD::merge(&mut accumulator, destination, origin);
+            }
+            jump *= 2;
+            sync_units();
+        }
+    } else {
+        // 25
```

Comment:
Maybe remove this 25 and the 2* UP

Reply:
Ha yes. They were notes for me to follow the algorithm :)